File: GFG POTD/021-August-2025/015-insert-interval.cpp
// https://www.geeksforgeeks.org/problems/insert-interval-1666733333/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // TC : O(N * log(N))
    // SC : O(N)
    vector<vector<int>> approach_1(vector<vector<int>> &intervals, vector<int> &newInterval) {
        intervals.push_back(newInterval);
        
        sort(begin(intervals), end(intervals));
        
        vector<vector<int>> result;
        for (vector<int> &interval : intervals) {
            if (result.empty() || interval[0] > result.back()[1]) {
                result.push_back(interval);
            } else {
                result.back()[1] = max(result.back()[1], interval[1]);
            }
        }
        return result;
    }
    
    // TC : O(N)
    // SC : O(N)
    vector<vector<int>> approach_2(vector<vector<int>> &intervals, vector<int> &newInterval) {
        int n = intervals.size();
        vector<vector<int>> result;
        
        // Step-1 : Left Part : Intervals which ends before newInterval start
        int i = 0;
        while (i < n && intervals[i][1] < newInterval[0]) {
            result.push_back(intervals[i]);
            i ++;
        }
        
        // Step-2 : Merge Part : Overlapping Interval
        while (i < n && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = min(intervals[i][0], newInterval[0]);
            newInterval[1] = max(intervals[i][1], newInterval[1]);
            i ++;
        }
        result.push_back(newInterval);
        
        // Step-3 : Right Part : Intervals which start after newintervals end
        while (i < n) {
            result.push_back(intervals[i]);
            i ++;
        }
        
        return result;
    }
    
    vector<vector<int>> insertInterval(vector<vector<int>> &intervals, vector<int> &newInterval) {
        // return approach_1(intervals, newInterval);
        return approach_2(intervals, newInterval);
    }
};

File: GFG POTD/021-August-2025/015-insert-interval.java
// https://www.geeksforgeeks.org/problems/insert-interval-1666733333/1




import java.util.*;

class Solution {
    public ArrayList<int[]> insertInterval(int[][] intervals, int[] newInterval) {
        // return approach_1(intervals, newInterval);
        return approach_2(intervals, newInterval);
    }
    
    // TC : O(N)
    // SC : O(N)
    private ArrayList<int[]> approach_2(int[][] intervals, int[] newIntervals) {
        int n = intervals.length;
        
        ArrayList<int[]> result = new ArrayList<>();
        int i = 0;
        
        // Step-1 : Left Part : Intervals which ends before the newIntervals start
        while (i < n && intervals[i][1] < newIntervals[0]) {
            result.add(intervals[i]);
            i ++;
        }
        
        // Step-2 : Merge Part : Merge overlapping intervals
        while (i < n && intervals[i][0] <= newIntervals[1]) {
            newIntervals[0] = Math.min(newIntervals[0], intervals[i][0]);
            newIntervals[1] = Math.max(newIntervals[1], intervals[i][1]);
            i ++;
        }
        result.add(newIntervals);
        
        // Step-3 : Right Part : Intervals which start after newInterval end
        while (i < n) {
            result.add(intervals[i]);
            i ++;
        }
        
        return result;
    }
    
    // TC : O(N * log(N))
    // SC : O(N)
    private ArrayList<int[]> approach_1(int[][] intervals, int[] newInterval) {
        ArrayList<int[]> intervalList = new ArrayList<>(Arrays.asList(intervals));
        intervalList.add(newInterval);
        
        intervalList.sort(Comparator.comparingInt(a -> a[0]));
        
        ArrayList<int[]> result = new ArrayList<>();
        
        for (int[] interval : intervalList) {
            if (result.isEmpty() || interval[0] > result.get(result.size() - 1)[1]) {
                result.add(interval);
            } else {
                result.get(result.size() - 1)[1] = Math.max(result.get(result.size() - 1)[1], interval[1]);
            }
        }
        
        return result;
    }
}